### PR TITLE
History redesign

### PR DESCRIPTION
**Pull request by xmrdsc**
History page rewritten.

Changes to what we currently have:

- Now called the 'transactions' page.
- Paginated results - works well for wallets with a lot of transactions (tested on wallet with 3700 txs).
  - Functionality to jump to a specific page in the pagination menu.
  - Pagination also works over filtered results.
- Configurable date display format: 'ago string' and 'YYYY-MM-hh:mm:ss'
- Blockheight now correctly reports 'transaction pool' when a TX in transaction pool
- When the transaction history changes due to incoming/outgoing funds, or simply needing to update the confirmations due to new blocks, the data will update, yet active filters and pagination will remain unchanged.

I did my best to create a pleasant UX experience. Something that is robust and just works. I like how I can quickly establish what my biggest transactions were, my oldest ones, or what I did last week in terms of transaction history. Now, this was already (mostly) possible with our old UX, however, this is nicer to operate and in addition adds new functionality. Try it out and let me know what you think.

Reference design: [Img](https://github.com/GBKS/monero-wallet-design/blob/master/screens/desktop-dark/transactions.png)

### CODE REVIEW

**Comment by selsta** [comment]
New design is looking really nice! :) Huge improvements IMO, also nice little features like jump to page.

Non-code feedback:

- Is it good to always have the first transaction open? I dislike how it always opens again.
- I noticed that the first two entered search characters seem to get ignored.
- Scrolling. The transactions seem to have their own scrolling behaviour, which doesn‘t work well on macOS.

**Comment by xmrdsc** [comment]
In addition some critique from Twitter which I agree on:

> no, why remove arrows? they are very easy to mentally figure out if TX is sent or received, same shaped circles suck.

Up/Down arrows were quite nice, what do you think @GBKS ?

**Comment by GBKS** [comment]
@xmrdsc maybe one of the things I tried to avoid is that we have an up/down arrow on the left for the transaction type, and another up/down arrow on the right for expanding/collapsing. Even if the arrows look a little different, it just seems like it could be confusing.

Maybe it's because I'm a designer, but I find that the green/orange color to indicate sent/received communicates very well. Others might not pay that much attention to color coding. We can totally change it though if this doesn't work for a lot of people and I can look into making sure we have unique icons that avoid the problem I mentioned above. What do you think?

**Comment by xmrdsc** [comment]
I might preference the old arrows, however, currently is not problematic in my opinion - ill follow your expertise.

**Comment by SamsungGalaxyPlayer** [comment]
I don't like the arrows since it's ambiguous which direction is sent and received anyway.

**Comment by selsta** [comment]
+1 for dots, the UI is already full of arrows.

**Comment by xmrdsc** [comment]
> Is it good to always have the first transaction open? I dislike how it always opens again.

Disabled that.

> I noticed that the first two entered search characters seem to get ignored.

It'll now start searching after 1 character. This was an optimization/performance thing - but seems to not matter much anyway.

> Scrolling. The transactions seem to have their own scrolling behaviour, which doesn‘t work well on macOS.

No idea on this one. I'm getting MacOS soon so I can test properly on OSX.

**Comment by xmrdsc** [comment]
Fixed MacOS scrolling by setting `interactive: false` on the listview.

File: pages/History.qml [hunk 1]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+                Layout.fillWidth: true
+                input.topPadding: 6 * scaleRatio
+                input.bottomPadding: 6 * scaleRatio
+                fontSize: 16 * scaleRatio
+                labelFontSize: 14 * scaleRatio
+                placeholderText: qsTr("Search...") + translationManager.emptyString
+                placeholderFontSize: 16 * scaleRatio
+                inputHeight: 34
+                onTextUpdated: {
+                    if(searchInput.text != null && searchInput.text.length >= 3){
```
Comment (selsta) [inline pages/History.qml]:
```suggestion
                    if(searchInput.text != null && searchInput.text.length >= 2){
```

> It'll now start searching after 1 character.

I think you forgot to change this.

File: LeftPanel.qml [hunk 2]
```diff
@@ -464,7 +464,7 @@ Rectangle {
                 id: historyButton
                 anchors.left: parent.left
                 anchors.right: parent.right
-                text: qsTr("History") + translationManager.emptyString
+                text: qsTr("Transactions") + translationManager.emptyString
```
[the remark this answers is not in the document]
Comment (xmrdsc) [reply LeftPanel.qml]: I agree. Lets do that in a later patch.

File: pages/History.qml (hunk 1 above)
Comment (xmrdsc) [reply pages/History.qml]: Fixed, thanks.

File: pages/History.qml [hunk 3]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
-      var sign = ""
-      if (total < 0) {
-        total = -total
-        sign = "-"
-      }
-      return count + qsTr(" selected: ") + sign + walletManager.displayAmount(total);
-    }
+    ColumnLayout {
+        id: pageRoot
+        anchors.topMargin: 20 * scaleRatio
```
[the remark this answers is not in the document]
Comment (xmrdsc) [reply pages/History.qml]: 👍

File: pages/History.qml [hunk 4]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+
+                    Rectangle {
                         anchors.top: parent.top
-                        anchors.left: parent.left
+                        anchors.topMargin: 24 * scaleRatio
+                        anchors.horizontalCenter: parent.horizontalCenter
+                        width: 10 * scaleRatio
+                        height: 10 * scaleRatio
+                        radius: 8 * scaleRatio
+                        color: isout ? "#d85a00" : "#2eb358"
```
Comment (mmbyday) [inline pages/History.qml]: https://github.com/monero-project/monero-gui/blob/d4d8ff54e6c94f82de7eb012d7b310addeedcbb9/components/HistoryTable.qml#L143

Can we highlight mining reward transactions like before, with a special color?

File: pages/History.qml [hunk 5]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+                                }
+                            }
+
+                            Image {
+                                anchors.top: parent.top
+                                anchors.left: txNoteText.right
+                                anchors.leftMargin: 12 * scaleRatio
+                                source: "../images/editIcon.png"
+                                opacity: 1
+                                width: 18
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                width: 18 * scaleRatio
```

File: pages/History.qml [hunk 6]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+                                height: 18
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                height: 18 * scaleRatio
```

File: pages/History.qml [hunk 7]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+
+                        Rectangle {
+                            color: "transparent"
+                            Layout.fillWidth: true
+                            Layout.preferredHeight: 20
+
+                            Text {
+                                font.family: MoneroComponents.Style.fontRegular.name
+                                font.pixelSize: 15 * scaleRatio
+                                text: "Transaction key"
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                text: qsTr("Transaction key")
```

File: pages/History.qml [hunk 8]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+
+                        Rectangle {
+                            color: "transparent"
+                            Layout.fillWidth: true
+                            Layout.preferredHeight: 20
+
+                            Text {
+                                font.family: MoneroComponents.Style.fontRegular.name
+                                font.pixelSize: 15 * scaleRatio
+                                text: "Transaction ID"
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                text: qsTr("Transaction ID")
```

File: pages/History.qml [hunk 9]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+
+                        Rectangle {
+                            color: "transparent"
+                            Layout.fillWidth: true
+                            Layout.preferredHeight: 20
+
+                            Text {
+                                font.family: MoneroComponents.Style.fontRegular.name
+                                font.pixelSize: 15 * scaleRatio
+                                text: "Description"
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                text: qsTr("Description")
```

File: pages/History.qml [hunk 10]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+
+                            Rectangle {
+                                color: "transparent"
+                                Layout.fillWidth: true
+                                Layout.preferredHeight: 20
+
+                                Text {
+                                    font.family: MoneroComponents.Style.fontRegular.name
+                                    font.pixelSize: 15 * scaleRatio
+                                    text: "Date"
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                    text: qsTr("Date")
```

File: pages/History.qml [hunk 11]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+
+                            Rectangle {
+                                color: "transparent"
+                                Layout.fillWidth: true
+                                Layout.preferredHeight: 20
+
+                                Text {
+                                    font.family: MoneroComponents.Style.fontRegular.name
+                                    font.pixelSize: 15 * scaleRatio
+                                    text: "Blockheight"
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                    text: qsTr("Blockheight")

File: pages/History.qml [hunk 12]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+
+                            Rectangle {
+                                color: "transparent"
+                                Layout.fillWidth: true
+                                Layout.preferredHeight: 20
+
+                                Text {
+                                    font.family: MoneroComponents.Style.fontRegular.name
+                                    font.pixelSize: 15 * scaleRatio
+                                    text: isout ? "Sent" : "Received"
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                    text: isout ? qsTr("Sent") : qsTr("Received")
```

File: pages/History.qml [hunk 13]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+
+                            Rectangle {
+                                color: "transparent"
+                                Layout.fillWidth: true
+                                Layout.preferredHeight: 20
+
+                                Text {
+                                    font.family: MoneroComponents.Style.fontRegular.name
+                                    font.pixelSize: 15 * scaleRatio
+                                    text: "Fee"
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                    text: qsTr("Fee")
```

File: pages/History.qml [hunk 14]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+
+                            Rectangle {
+                                color: "transparent"
+                                Layout.fillWidth: true
+                                Layout.preferredHeight: 20
+
+                                Text {
+                                    font.family: MoneroComponents.Style.fontRegular.name
+                                    font.pixelSize: 15 * scaleRatio
+                                    text: "Confirmations"
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                    text: qsTr("Confirmations")
```

File: pages/History.qml [hunk 15]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+                        Rectangle {
+                            visible: isout
+                            color: "transparent"
+                            Layout.fillWidth: true
+                            Layout.preferredHeight: 20
+
+                            Text {
+                                font.family: MoneroComponents.Style.fontRegular.name
+                                font.pixelSize: 15 * scaleRatio
+                                text: "Address sent to"
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                                text: "Address sent to"
```
```suggestion
                                text: qsTr("Address sent to")
```

File: pages/History.qml [hunk 16]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+                        source: "../images/whiteDropIndicator.png"
+                        rotation: delegate.collapsed ? 180 : 0
                     }
                 }
+
+                Rectangle {
+                    anchors.left: parent.left
+                    anchors.right: parent.right
+                    anchors.top: parent.top
+                    height: 1
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                    height: 1 * scaleRatio
```

File: pages/History.qml [hunk 17]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+                    color: "#2b2b2b"
+                }
+
+                Rectangle {
+                    anchors.left: parent.left
+                    anchors.right: parent.right
+                    anchors.top: parent.bottom
+                    height: 1
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
                    height: 1 * scaleRatio
```

File: pages/History.qml [hunk 18]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+        for (var i = 0; i < txs.length; i++){
+            txListViewModel.append(txs[i]);
+        }
+
+        root.updateHistoryStatusMessage();
+
+        // determine pagination button states
+        var count = txData.length;
+        if(count <= root.txMax) {
+            paginationPrev.enabled = false
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
        informationPopup.title = qsTr("Transaction details");
```

File: pages/History.qml [hunk 19]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+                            visible: isout
+                            color: "transparent"
+                            Layout.fillWidth: true
+                            Layout.preferredHeight: 20
+
+                            Text {
+                                font.family: MoneroComponents.Style.fontRegular.name
+                                font.pixelSize: 15 * scaleRatio
+                                text: {
+
```
Comment (mmbyday) [inline pages/History.qml]:
```suggestion
```

File: pages/History.qml (hunk 18 above)
Comment (mmbyday) [reply pages/History.qml]: Weird glitch here on github. Should be line 1496 needs a qsTr.

File: pages/History.qml [hunk 20]
```diff
@@ -34,342 +34,1519 @@ import moneroComponents.WalletManager 1.0
...
+                            Rectangle {
+                                color: "transparent"
+                                Layout.fillWidth: true
+                                Layout.preferredHeight: 20
+
+                                Text {
+                                    font.family: MoneroComponents.Style.fontRegular.name
+                                    font.pixelSize: 14 * scaleRatio
+                                    text: blockheight > 0 ? blockheight : qsTr('Transaction Pool');
```
[the remark this answers is not in the document]
Comment (xmrdsc) [reply pages/History.qml]: :+1:

File: pages/History.qml (hunk 4 above)
Comment (xmrdsc) [reply pages/History.qml]: Sorry for not migrating this cool feature. Seems I forgot. Here is how it looks like with mining:

File: pages/History.qml (hunk 5 above)
Comment (xmrdsc) [reply pages/History.qml]: @selsta and I have been talking, we're moving away from scaleRatio. It has absolutely no use right now. Instead, people should use `QT_SCALE_FACTOR`.

File: pages/History.qml (hunk 6 above)
Comment (xmrdsc) [reply pages/History.qml]: See reply above.

File: pages/History.qml (hunk 7 above)
Comment (xmrdsc) [reply pages/History.qml]: whoops :)

File: pages/History.qml (hunk 8 above)
Comment (xmrdsc) [reply pages/History.qml]: whoops :)

File: pages/History.qml (hunk 10 above)
Comment (xmrdsc) [reply pages/History.qml]: whoops

File: pages/History.qml (hunk 12 above)
Comment (xmrdsc) [reply pages/History.qml]: wh00pz x5

File: pages/History.qml (hunk 13 above)
Comment (xmrdsc) [reply pages/History.qml]: :+1:

File: pages/History.qml (hunk 14 above)
Comment (xmrdsc) [reply pages/History.qml]: :+1:

File: pages/History.qml (hunk 15 above)
Comment (xmrdsc) [reply pages/History.qml]: :+1:

File: pages/History.qml (hunk 16 above)
Comment (xmrdsc) [reply pages/History.qml]: See reply above.

File: pages/History.qml (hunk 17 above)
Comment (xmrdsc) [reply pages/History.qml]: See reply above.

File: pages/History.qml (hunk 18 above)
Comment (xmrdsc) [reply pages/History.qml]: :+1:

File: pages/History.qml (hunk 19 above)
Comment (xmrdsc) [reply pages/History.qml]: :+1:

**Comment by xmrdsc** [comment]
Thanks @selsta and @mmbyday for the extensive reviews. Looks done to me.

**Pull request merged**